src/compute-client/src/controller/replicated.rs

```diff
@@ -77,27 +86,89 @@ where
+    /// Specialize a command for the given `Replica` and `ReplicaId`.
+    ///
+    /// Most `ComputeCommand`s are independent of the target replica, but some
+    /// contain replica-specific fields that must be adjusted before sending.
+    fn specialize_command<T>(&self, command: &mut ComputeCommand<T>) {
+        // Set new replica ID and obtain set the sinked logs specific to this replica
+        if let ComputeCommand::CreateInstance(config) = command {
+            config.replica_id = self.replica_id;
+            config.logging = self.logging_config.clone();
+        }
+
+        if let ComputeCommand::CreateTimely(comm_config) = command {
+            *comm_config = self.comm_config.clone();
+        }
+    }
+}
+
 async fn run_replica_core<T>(
     ReplicaTaskConfig {
+        instance_id,
         replica_id,
-        addrs,
+        location,
+        logging_config,
         build_info,
         mut command_rx,
         response_tx,
+        orchestrator,
     }: ReplicaTaskConfig<T>,
 ) -> Result<(), anyhow::Error>
 where
     T: Timestamp + Lattice,
     ComputeGrpcClient: ComputeClient<T>,
 {
-    let mut client = Retry::default()
+    let (mut client, comm_config) = Retry::default()
         .clamp_backoff(Duration::from_secs(32))
         .retry_async(|state| {
-            let addrs = addrs.clone();
             let version = build_info.semver_version();
+            let location = location.clone();
+            let orchestrator = orchestrator.clone();
+
             async move {
+                let (addrs, comm_config) = match location {
+                    ComputeReplicaLocation::Remote {
+                        addrs,
+                        compute_addrs,
+                        workers,
+                    } => {
+                        let addrs = addrs.into_iter().collect();
+                        let comm_config = CommunicationConfig {
+                            workers: workers.get(),
+                            process: 0,
+                            addresses: compute_addrs.into_iter().collect(),
+                        };
+                        (addrs, comm_config)
+                    }
+                    ComputeReplicaLocation::Managed {
+                        allocation,
+                        availability_zone,
+                        ..
+                    } => {
+                        let service = orchestrator
+                            .ensure_replica(instance_id, replica_id, allocation, availability_zone)
+                            .await?;
+
+                        let addrs = service.addresses("controller");
+                        let comm_config = CommunicationConfig {
+                            workers: allocation.workers.get(),
+                            process: 0,
+                            addresses: service.addresses("compute"),
+                        };
+                        (addrs, comm_config)
+                    }
+                };
```

Comment:
Can we move the orchestrator call in a separate retry loop? If the connection to the replica fails and needs to be retried, there is no reason to also invoke the orchestrator again. I would also be fine with not retrying the orchestrator call at all and just panicking, but retrying is probably the safer default.

Reply:
see above ^